src/riak_ensemble_peer.erl

```diff
@@ -1884,22 +1902,29 @@ handle_sync_event(tree_info, _From, StateName, State=#state{tree_trust=Trust,
 handle_sync_event({debug_local_get, Key}, From, StateName, State) ->
     State2 = do_local_get(From, Key, State),
     {next_state, StateName, State2};
+handle_sync_event(get_watchers, _From, StateName, State) ->
+    {reply, State#state.watchers, StateName, State};
 handle_sync_event(_Event, _From, StateName, State) ->
     Reply = ok,
     {reply, Reply, StateName, State}.
 
 %% -spec handle_info(_, atom(), state()) -> next_state().
-handle_info({'DOWN', Ref, _, Pid, Reason}, StateName,
-  #state{mod=Mod, modstate=ModState}=State) ->
-    case Mod:handle_down(Ref, Pid, Reason, ModState) of
+handle_info({'DOWN', MRef, _, Pid, Reason}, StateName, State) ->
+    #state{mod=Mod, modstate=ModState, watchers=Watchers} = State,
+    case lists:keytake(Pid, 1, Watchers) of
```

Comment:
Yeah funny you would suggest that, because I did also consider splitting the nested case out into separate functions. I thought it was still pretty understandable as is, so I decided to leave all the code in one place, but in hindsight I am probably biased since I was writing the code and not just reading it.

If it seemed obtuse to you then I'm happy to refactor it - will update the PR shortly.